lncore/addr.go

```diff
@@ -0,0 +1,319 @@
+package lncore
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"strconv"
+	"encoding/hex"
+
+	"github.com/mit-dci/lit/bech32"
+	"github.com/mit-dci/lit/crypto/fastsha256"
+)
+
+/*
+ * Ok so there's a lot of stuff going on here.  Lit natively supports addresses
+ * that are just the pkh of the pubkey encoded in bech32.  And that's pretty
+ * cool, you look up the IP+port in a tracker (or DHT) and connected with
+ * Noise_XX, then verify that the pubkey they provide you matches the pkh you
+ * already have.  Also you can provide IP+port if you want.
+ *
+ * But in mainstream BOLT nodes they use the full pubkey hex-encoded.  They have
+ * a DHT where you look up the IP+port (if not provided) and connect with
+ * Noise_XK, which assumes you already have the pubkey (since you do) and that's
+ * all fine.
+ *
+ * This tries to unify all of them with formats like this under 1 parser:
+ *
+ *      <bech32 pkh>[@<ip>[:<port>]
+ * or
+ *
+ *      <hex of bech32 pubkey>[@<ip>[:<port>]
+ *
+ * or fully
+ *
+ *      <bech32 pkh>:<pubkey>[@<ip>:<port>]
+ */
+
+// LnAddr is just a bech32-encoded pubkey.
+// TODO Move this to another package so it's more obviously not *just* IO-related.
+type LnAddr string
+
+// ToString returns the LnAddr as a string.  Right now it just unwraps it but it
+// might do something more eventually.
+func (lnaddr LnAddr) ToString() string {
+	return string(lnaddr)
+}
+
+// LnDefaultPort is from BOLT1.
+const LnDefaultPort = 9735
+
+// PkhBech32Prefix if for addresses lit already uses.
+const PkhBech32Prefix = "ln"
+
+// PubkeyBech32Prefix is for encoding the full pubkey in bech32.
+const PubkeyBech32Prefix = "lnpk"
```

Comment:
Why would we need this though, can't we parse with just the length of the resultant `bech32` string?

Reply:
Yeah that's true, I suppose I wanted to make the difference more obvious.  Especially since technically `ln1dfasdfasdf:lnpk1asdfsadfasdf` is a valid address, it'd look a little weird to me if it was `ln1:ln1`, and it'd be less obvious for the average joe to know how to pick out the "short" version of their address, which we want to make the pkh.

Reply:
Sounds good thanks! Although I wonder why people would want to use bech32 pks (since lit right now supports only bech32 pkh and BOLT supports hex pks). Would be nice if some implementations decide to change though, so having this does make sense.

Reply:
I mean, it doesn't *really* matter.  They get a little bit of error correction (maybe), but since most of the time addresses are going to be copy/pasted I don't think it matters.  I added it to quell arguments about "well hex might be bad but bech32 is better" since this is demonstrably not that much shorter and not worth the extra effort.